proposals/0025-shader-execution-reordering.md

```diff
@@ -0,0 +1,1519 @@
+
+For performance reasons, it is crucial that the DXIL-generating compiler does
+not move non-uniform resource access across reorder points in general, and across
+`ReorderThread` in particular. It should be assumed that the shader will perform
+the access where coherence is maximized.
+
+---
+
+### Divergence behavior
+
+An implementation may only reorder threads whose control flow arrives at a
+reorder point. Threads that do not arrive at a reorder point are guaranteed to
+not migrate, even if neighboring threads in the same wave do migrate.
+
+From the physical wave's point of view, an implementation may remove, replace,
+or do nothing with threads that arrive at a reorder point. It is also legal for
+an implementation to replace threads that were inactive at the start of the
+wave or those that exited the shader during execution, as long as any remaining
+thread reach a reorder point. This may indirectly affect threads that
+conditionally did not invoke a reorder point, as illustrated in the following
+code snippet:
+
+```C++
+int MyFunc(int coherenceCoord)
+{
+    int A = WaveActiveBallot(true);
+    if (WaveIsFirstLane())
+        ReorderThread(coherenceCoord, 32);
+    int B = WaveActiveBallot(true);
+    return A - B;
+}
+```
+
+In this example, a number of different things could happen:
+- If the implementation does not honor `ReorderThread` at all, the function
+will most likely return zero, as the set of threads before and after the
+conditional reorder would be the same.
+- If the implementation reorders threads invoking `ReorderThread` but does not
+replace them, B will likely be less than A for threads not invoking
+`ReorderThread`, while the reordered threads will likely resume execution with
+a newly formed full wave, thereby obtaining `A <= B`.
+- If the implementation replaces threads in a wave, the threads not
+participating in the reorder may possibly be joined by more threads than were
+removed from the wave, again observing `A <= B`.
+
+---
+
+### Memory coherence and visibility
+
+Due to the existence of non-coherent caches on most modern GPUs, an
+application must take special care when communicating information across
+reorder points through memory (UAVs).
+
+This proposal includes a new coherence scope for reorder points. The scope
+is limited to communication within the same dispatch index. Specifically,
+if UAV stores or atomics are performed on one side of a reorder point, and
+on the other side the data is read via non-atomic UAV reads, the following
+is required:
+- The UAV must be declared `[reordercoherent]`.
+- The UAV writer must issue a `Barrier(UAV_MEMORY, REORDER_SCOPE)` between
+the write and the reorder point.
+
+When communicating both between threads and across reorder points, global coherency
+can be utilized:
+- The UAV must be declared `[globallycoherent]`.
+- The UAV writer must issue a `DeviceMemoryBarrier` between the write and the
+reorder point.
```

Comment:
From my point of view, the `DeviceMemoryBarrier` call makes sure that all writes have completed before the reorder point gets to move the work to a different thread. When reading after the reorder point, there's no ordering problem. The only risk is that stale data is read from a local cache. That is mitigated by the `globallycoherent` attribute, causing the local cache to be bypassed.

Reply:
Let's separate the `reordercoherent` and `globallycoherent` discussions. I think `globallycoherent` in DXR semantics are a bit problematic because of the synchronization limitations, but that's not the focus of this change.

For `reordercoherent`, are you saying that you agree that not requiring a barrier between the reorder point and the load *does penalize* some implementations, but we accept that in favour of a simpler API that is closer to what e.g. WG require?

Reply:
What I'm trying to say is that we are not trying to add acquire and release semantics to HLSL with this spec proposal.

What you are suggesting is calling `Barrier` both before and after the reorder point. First of all, note that you would have to have a barrier after each and every reorder point that wants to read the data after a single write since each reorder point will be a new thread. So a single write can result in a many required barriers, which may be difficult for the user. Second, there is no distinction between acquire and release with `Barrier` which means that all possible implementations are harmed (e.g., ordering + cache flush at both sides). Finally, it is not clear that there is any case where a high-frequency cache flush would be better than adopting the `globallycoherent` semantics of bypassing local caches, which all implementations already supports.

Reply:
> What I'm trying to say is that we are not trying to add acquire and release semantics to HLSL with this spec proposal.

Agreed.

> 
> What you are suggesting is calling `Barrier` both before and after the reorder point. First of all, note that you would have to have a barrier after each and every reorder point that wants to read the data after a single write since each reorder point will be a new thread. 

Nit: Let's not mix the thread terminology (as in the `globallycoherent` discussion) with the reorder terminology. We can say that at reorder points, `reordercoherent` synchronization is necessary. If we want to give it a name, maybe "execution locus" is a better term instead of "thread"?

> So a single write can result in a many required barriers, which may be difficult for the user. 

Agreed.

> Second, there is no distinction between acquire and release with `Barrier` which means that all possible implementations are harmed (e.g., ordering + cache flush at both sides). 

In the worst case, putting a `reordercoherent` barrier immediately after every reorder point (CallShader/TraceRay/ReportHit/HitObject ops / shader entry) should be equivalent to the current proposed behavior without the barrier on the reading side. Because it immediately follows a reorder point, it is clear that only reads can be affected, and if the implementation uses the bypass strategy nothing needs to be done. 
An implementation that doesn't use cache bypass would have to add exactly all these barriers, and the explicit one just makes that explicit.

> Finally, it is not clear that there is any case where a high-frequency cache flush would be better than adopting the `globallycoherent` semantics of bypassing local caches, which all implementations already supports.

Agreed that this is totally unclear in general, but I think when designing a specification that is supposed to be used for an extended lifetime, including on not yet existing hardware, one should be cautious to unnecessarily restrict the freedom of possible implementations.

To be clear, I am not demanding to change it to require the barrier on the reading side, I'm rather pointing out a potential issue. If there is consensus that the single-barrier solution is preferable, I'm totally fine with that.

Reply:
We re-discussed internally and are fine with the proposed solution of only requiring a barrier on the write side. It may penalize some implementations, but it's really a very special corner case where it makes a difference, and it's questionable whether it's worth the effort optimizing for that.